0000-deb-repo-sync-gpg-check.md

```diff
@@ -0,0 +1,178 @@
+
+# Detailed design
+
+## Database changes
+
+A new table is needed to store the association of a GPG key to a repository:
+```
+rhnContentSourceGpg
+====================
+content_source_id   number not null foreign key rhnContentSource(id)
+gpg_key_id          number not null foreign key rhnCryptoKey(id)
+created             timestamp
+modified            timestamp
+
+```
+
+## Repo signature checking
+
+The GPG verification should be optional but a warning should be shown to the the user in case he's not providing information about the GPG key used to verify the metadata.
+
+### Repo metadata location and signatures
+
+_Note: this section is only informative_
+
+Debian/Ubuntu repos use GPG signing to ensure the integrity of the `Release` file. The signature can be either in a separate file `Release.gpg` or inline in a file called `InRelease`. If both files are present the `InRelease` file is preferred.
+
+Typically the structure of a repo is like this:
+```
+http://mirror.example.com/ubuntu  <- $ARCHIVE_ROOT
+|
++- dists/bionic                   <- Suite or Codename
+   |
+   +- main                        <- $COMPNENT directories
+      |
+      +- binary-amd64
+         |
+         +- Packages.gz           <- Packages indices
+         +- Packages.xz
+         . . .
+      +- binary-i386
+      . . .
+   +- universe
+   . . .
+   +- Release                     <- Plain file
+   +- InRelease                   <- Release file + inline signature
+   +- Release.gpg                 <- Separate signature
+
+```
+
+E.g. for official Ubuntu repos the URL of these files:
+```
+http://mirror.example.com/ubuntu/dists/bionic-updates/InRelease
+http://mirror.example.com/ubuntu/dists/bionic-updates/Release.gpg
+```
+
+There's also an alternative layout for Debian repos called "flat repos". In this case the repo directory containins both `Release*` and `Packages*` files. E.g.:
+```
+https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
+```
+
+_Note #1:_
+
+In Uyuni the URL of a deb repository must point to the directory containing the binary packages for a specific architecture. E.g. for Ubuntu 18.04 updates repo is something like:
+```
+http://mirror.example.com/ubuntu/dists/bionic-updates/main/binary-amd64/
+```
+or in case of a flat repository to the directory containing both `Release*` and `Packages*` files:
+```
+https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
+```
+
+_Note #2:_
+
+Debian repos don't have a standard location for the public GPG keys used to verify the metadata. Sometimes the key is located in the root of a flat repo but the location is not standardized.
+
+### GPG keys management
+
+Uyuni ships the SUSE and OpenSUSE keys in the package `uyuni-build-keys`. The keys for Ubuntu and Debian should be shipped in this package too.
```

Comment:
What if a user do not want to trust such a key? I would suggest to keep the automatically trusted keys very limited.
An option could be to create sub-packages and the admin can decide which keys he want to trust by install them or not.

Reply:
Changed to ship Debian and Ubuntu keys in separate packages.